File: src/main/java/org/nest/nestml/_visitor/NumericLiteralVisitor.java
package org.nest.nestml._visitor;

import de.monticore.literals.literals._ast.ASTDoubleLiteral;
import de.monticore.literals.literals._ast.ASTIntLiteral;
import de.monticore.symboltable.Scope;
import org.nest.nestml._ast.ASTExpr;
import org.nest.nestml._symboltable.typechecking.Either;
import org.nest.nestml._symboltable.symbols.VariableSymbol;
import org.nest.nestml._symboltable.unitrepresentation.UnitTranslator;

import java.util.Optional;

import static org.nest.nestml._symboltable.predefined.PredefinedTypes.*;

/**
 * @author ptraeder
 */
public class NumericLiteralVisitor implements NESTMLVisitor{
  UnitTranslator unitTranslator = new UnitTranslator();
  @Override
  public void visit(ASTExpr expr) {
    //if variable is also set in this expression, the var type overrides the literal
    if(expr.getVariable().isPresent()){
      final Scope scope = expr.getEnclosingScope().get();
      final String varName = expr.getVariable().get().toString();
      final Optional<VariableSymbol> variableSymbol = scope.resolve(varName, VariableSymbol.KIND);
      expr.setType(Either.value(variableSymbol.get().getType()));
      return;
    }

    if (expr.getNumericLiteral().get() instanceof ASTDoubleLiteral) {
      expr.setType(Either.value(getRealType()));
      return;
    }
    else if (expr.getNumericLiteral().get() instanceof ASTIntLiteral) {
      expr.setType(Either.value(getIntegerType()));
      return;
    }
  }
}

File: pynestml/src/main/python/org/nestml/visitor/expression_visitor/NumericLiteralVisitor.py
#
# NumericLiteralVisitor.py
#
# This file is part of NEST.
#
# Copyright (C) 2004 The NEST Initiative
#
# NEST is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# NEST is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NEST.  If not, see <http://www.gnu.org/licenses/>.

"""
simpleexpression : (INTEGER|FLOAT) (variable)?
"""
from pynestml.src.main.python.org.nestml.symbol_table.predefined.PredefinedTypes import PredefinedTypes
from pynestml.src.main.python.org.nestml.symbol_table.symbols.Symbol import SymbolKind
from pynestml.src.main.python.org.nestml.visitor.NESTMLVisitor import NESTMLVisitor
from pynestml.src.main.python.org.nestml.visitor.expression_visitor.Either import Either


class NumericLiteralVisitor(NESTMLVisitor):

    def visitSimpleExpression(self, _expr=None):
        assert _expr.getScope() is not None, "Run symboltable creator."
        #if variable is also set in this expression, the var type overrides the literal
        if _expr.getVariable() is not None:
            scope = _expr.getScope()
            varName = _expr.getVariable().getName()
            variableSymbolResolve = scope.resolveToSymbol(varName, SymbolKind.VARIABLE)
            _expr.setTypeEither(Either.value(variableSymbolResolve.getTypeSymbol()))
            return

        if _expr.getNumericLiteral() is not None and isinstance(_expr.getNumericLiteral,float):
            _expr.setTypeEither(Either.value(PredefinedTypes.getRealType()))
            return

        elif _expr.getNumericLiteral() is not None and isinstance(_expr.getNumericLiteral,int):
            _expr.setTypeEither(Either.value(PredefinedTypes.getIntegerType()))
            return